stl/src/vector_algorithms.cpp

```diff
@@ -260,6 +330,22 @@ __declspec(noalias) void __cdecl __std_reverse_trivially_swappable_4(void* _Firs
 
 __declspec(noalias) void __cdecl __std_reverse_trivially_swappable_8(void* _First, void* _Last) noexcept {
 #if !defined(_M_ARM64EC)
+    if (_Byte_length(_First, _Last) >= 128 && _bittest(&__isa_enabled, __ISA_AVAILABLE_AVX512)) {
+        const __m512i _Reverse_8x64bit_avx512 = _mm512_set_epi64(0, 1, 2, 3, 4, 5, 6, 7);
+        const void* _Stop_at                  = _First;
+        _Advance_bytes(_Stop_at, _Byte_length(_First, _Last) >> 7 << 6);
+        do {
+            _Advance_bytes(_Last, -64);
+            const __m512i _Left           = _mm512_loadu_si512(static_cast<__m512i*>(_First));
+            const __m512i _Right          = _mm512_loadu_si512(static_cast<__m512i*>(_Last));
+            const __m512i _Left_reversed  = _mm512_permutexvar_epi64(_Reverse_8x64bit_avx512, _Left);
+            const __m512i _Right_reversed = _mm512_permutexvar_epi64(_Reverse_8x64bit_avx512, _Right);
+            _mm512_storeu_si512(static_cast<__m512i*>(_First), _Right_reversed);
+            _mm512_storeu_si512(static_cast<__m512i*>(_Last), _Left_reversed);
+            _Advance_bytes(_First, 64);
+        } while (_First != _Stop_at);
+    }

```

Comment:
Maybe `else` here? if 512-bit version executed, then 256-bit version would execute either zero times or once, so it may be better to save on 256-b it version warming up and skip to 128-bit version instead (which would then execute zero to three times) (ditto other occurrences).

Reply:
I think it would be better to decide based on profiling results, instead of assuming that the 256-bit vector instructions will have a warm-up cost (they might, but let's confirm).

Reply:
I observed that in a similar situation in the product I'm working on, omitting some bit width caused better performance. Not sure if it is warming up, smaller function size, or unpredictable branch.
Definitely agree that it either should be motivated by profiling, or kept as is.

Reply:
I adopted the benchmark to run up to two AVX2 loops against a direct call of the tail loop. So to say close the least amount of data such that the warm-up phase may happen. The benchmark ran on the azure machine mentioned. I uploaded the benchmark to [this gist](https://gist.github.com/michael-herwig/2cabd4696e3c55cdda73d173312cb2d3), I also added instructions to zero the contents of all AVX2  XMM and YMM registers and ran one too small instance of the AVX2 algorithm to try to prevent caching and branch prediction. Please note time specifications are now in CPU cycles using the `__rdtsc()` intrinsic.

<details>
<summary>benchmark results</summary></br>

|Suite         |Benchmark|bytes  |D  |S  |Total cycles|Result|cylces/op|Baseline|
|--------------|---------|---|---|---|--------|------|-----|--------|
|swap          |baseline |*  |32 |4096|80      |0     |2    |1.000   |
|swap          |avx2     |   |32 |4096|26      |0     |0    |0.325   |
|reverse_1     |baseline |*  |8  |4096|26      |0     |3    |1.000   |
|reverse_1     |baseline |*  |16 |4096|24      |0     |1    |1.000   |
|reverse_1     |baseline |*  |32 |4096|46      |0     |1    |1.000   |
|reverse_1     |baseline |*  |64 |4096|70      |0     |1    |1.000   |
|reverse_1     |avx2     |   |8  |4096|26      |0     |3    |1.000   |
|reverse_1     |avx2     |   |16 |4096|30      |0     |1    |1.250   |
|reverse_1     |avx2     |   |32 |4096|38      |0     |1    |0.826   |
|reverse_1     |avx2     |   |64 |4096|28      |0     |0    |0.400   |
|reverse_2     |baseline |*  |8  |4096|22      |0     |2    |1.000   |
|reverse_2     |baseline |*  |16 |4096|18      |0     |1    |1.000   |
|reverse_2     |baseline |*  |32 |4096|26      |0     |0    |1.000   |
|reverse_2     |baseline |*  |64 |4096|56      |0     |0    |1.000   |
|reverse_2     |avx2     |   |8  |4096|18      |0     |2    |0.818   |
|reverse_2     |avx2     |   |16 |4096|18      |0     |1    |1.000   |
|reverse_2     |avx2     |   |32 |4096|32      |0     |1    |1.231   |
|reverse_2     |avx2     |   |64 |4096|26      |0     |0    |0.464   |
|reverse_4     |baseline |*  |8  |4096|22      |0     |2    |1.000   |
|reverse_4     |baseline |*  |16 |4096|24      |0     |1    |1.000   |
|reverse_4     |baseline |*  |32 |4096|20      |0     |0    |1.000   |
|reverse_4     |baseline |*  |64 |4096|22      |0     |0    |1.000   |
|reverse_4     |avx2     |   |8  |4096|24      |0     |3    |1.091   |
|reverse_4     |avx2     |   |16 |4096|20      |0     |1    |0.833   |
|reverse_4     |avx2     |   |32 |4096|22      |0     |0    |1.100   |
|reverse_4     |avx2     |   |64 |4096|26      |0     |0    |1.182   |
|reverse_8     |baseline |*  |8  |4096|20      |0     |2    |1.000   |
|reverse_8     |baseline |*  |16 |4096|22      |0     |1    |1.000   |
|reverse_8     |baseline |*  |32 |4096|20      |0     |0    |1.000   |
|reverse_8     |baseline |*  |64 |4096|20      |0     |0    |1.000   |
|reverse_8     |avx2     |   |8  |4096|22      |0     |2    |1.100   |
|reverse_8     |avx2     |   |16 |4096|22      |0     |1    |1.000   |
|reverse_8     |avx2     |   |32 |4096|22      |0     |0    |1.100   |
|reverse_8     |avx2     |   |64 |4096|24      |0     |0    |1.200   |
|reverse_copy_1|baseline |*  |8  |4096|28      |0     |3    |1.000   |
|reverse_copy_1|baseline |*  |16 |4096|36      |0     |2    |1.000   |
|reverse_copy_1|baseline |*  |32 |4096|52      |0     |1    |1.000   |
|reverse_copy_1|baseline |*  |64 |4096|84      |0     |1    |1.000   |
|reverse_copy_1|avx2     |   |8  |4096|34      |0     |4    |1.214   |
|reverse_copy_1|avx2     |   |16 |4096|46      |0     |2    |1.278   |
|reverse_copy_1|avx2     |   |32 |4096|22      |0     |0    |0.423   |
|reverse_copy_1|avx2     |   |64 |4096|18      |0     |0    |0.214   |
|reverse_copy_2|baseline |*  |8  |4096|20      |0     |2    |1.000   |
|reverse_copy_2|baseline |*  |16 |4096|28      |0     |1    |1.000   |
|reverse_copy_2|baseline |*  |32 |4096|36      |0     |1    |1.000   |
|reverse_copy_2|baseline |*  |64 |4096|52      |0     |0    |1.000   |
|reverse_copy_2|avx2     |   |8  |4096|20      |0     |2    |1.000   |
|reverse_copy_2|avx2     |   |16 |4096|30      |0     |1    |1.071   |
|reverse_copy_2|avx2     |   |32 |4096|20      |0     |0    |0.556   |
|reverse_copy_2|avx2     |   |64 |4096|26      |0     |0    |0.500   |
|reverse_copy_4|baseline |*  |8  |4096|18      |0     |2    |1.000   |
|reverse_copy_4|baseline |*  |16 |4096|24      |0     |1    |1.000   |
|reverse_copy_4|baseline |*  |32 |4096|26      |0     |0    |1.000   |
|reverse_copy_4|baseline |*  |64 |4096|36      |0     |0    |1.000   |
|reverse_copy_4|avx2     |   |8  |4096|22      |0     |2    |1.222   |
|reverse_copy_4|avx2     |   |16 |4096|26      |0     |1    |1.083   |
|reverse_copy_4|avx2     |   |32 |4096|22      |0     |0    |0.846   |
|reverse_copy_4|avx2     |   |64 |4096|22      |0     |0    |0.611   |
|reverse_copy_8|baseline |*  |8  |4096|22      |0     |2    |1.000   |
|reverse_copy_8|baseline |*  |16 |4096|20      |0     |1    |1.000   |
|reverse_copy_8|baseline |*  |32 |4096|24      |0     |0    |1.000   |
|reverse_copy_8|baseline |*  |64 |4096|22      |0     |0    |1.000   |
|reverse_copy_8|avx2     |   |8  |4096|22      |0     |2    |1.000   |
|reverse_copy_8|avx2     |   |16 |4096|20      |0     |1    |1.000   |
|reverse_copy_8|avx2     |   |32 |4096|18      |0     |0    |0.750   |
|reverse_copy_8|avx2     |   |64 |4096|24      |0     |0    |1.091   |
</details>

Reply:
The benchmark does not indicate any significant performance gain or drop, and results vary for similar algorithms. One may profile this in a real-world usage case, but it seems better to leave it as is.